metricbeat/helper/prometheus/prometheus.go

```diff
@@ -76,110 +86,224 @@ func (p *prometheus) GetFamilies() ([]*dto.MetricFamily, error) {
 	}
 
 	families := []*dto.MetricFamily{}
-	for {
-		mf := &dto.MetricFamily{}
-		err = decoder.Decode(mf)
-		if err != nil {
-			if err == io.EOF {
-				break
+	var wg sync.WaitGroup
+	var decoderMutex = &sync.Mutex{}
+	var familiesMutex = &sync.Mutex{}
+	var done = false
+	var count = 0
+	for !done && count < maxGetFamiliesSubFuns {
+		wg.Add(1)
+		count++
+		go func(id int) {
+			for !done {
+				mf := &dto.MetricFamily{}
+				decoderMutex.Lock()
+				err = decoder.Decode(mf)
+				decoderMutex.Unlock()
+				if err != nil {
+					if err == io.EOF {
+						done = true
```

Comment:
you can just return or break here and wouldn't it be the same? then you won't have the 'else' clause, so the code will be less indented and easier to read.  also, i can see why you need to set the 'done' variable in line 105. i guess you need it for line 94, but i don't think you need it for line 98 - can't this just be a no-condition-for-loop that just terminates when you hit EOF and then you return from function after seetting your 'done' variable in line 105?